### PR TITLE
gitlab_hook: use None for non-existent attr in gitlab API response

### PR DESCRIPTION
**Pull request by crashwind**
##### SUMMARY
there is a problem when webhook token is defined with ansible, because gitlab API doesn`t return 'token' attribute

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
gitlab_hook

##### ADDITIONAL INFORMATION

```
  - name: my/project crowdin webhook
    become: no
    local_action:
      module: community.general.gitlab_hook
      api_url: "{{ gitlab__api_url }}"
      api_token: "{{ vault.gitlab__api_token }}"
      state: present
      project: "my/project"
      hook_url: "https://crowdin.com/hooks/gitlab_server"
      push_events: yes
      hook_validate_certs: yes
      token: "{{ vault.gitlab__hook.crowdin_token }}"
```

error message
```
Traceback (most recent call last):
  File \"<stdin>\", line 102, in <module>
  File \"<stdin>\", line 94, in _ansiballz_main
  File \"<stdin>\", line 40, in invoke_module
  File \"/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/runpy.py\", line 207, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File \"/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/runpy.py\", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File \"/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/runpy.py\", line 87, in _run_code
    exec(code, run_globals)
  File \"/var/folders/l1/d5msh66s31ngsr9gdjc4kzx80000gn/T/ansible_community.general.gitlab_hook_payload_2yfbdmwa/ansible_community.general.gitlab_hook_payload.zip/ansible_collections/community/general/plugins/modules/source_control/gitlab/gitlab_hook.py\", line 390, in <module>
  File \"/var/folders/l1/d5msh66s31ngsr9gdjc4kzx80000gn/T/ansible_community.general.gitlab_hook_payload_2yfbdmwa/ansible_community.general.gitlab_hook_payload.zip/ansible_collections/community/general/plugins/modules/source_control/gitlab/gitlab_hook.py\", line 370, in main
  File \"/var/folders/l1/d5msh66s31ngsr9gdjc4kzx80000gn/T/ansible_community.general.gitlab_hook_payload_2yfbdmwa/ansible_community.general.gitlab_hook_payload.zip/ansible_collections/community/general/plugins/modules/source_control/gitlab/gitlab_hook.py\", line 212, in create_or_update_hook
  File \"/var/folders/l1/d5msh66s31ngsr9gdjc4kzx80000gn/T/ansible_community.general.gitlab_hook_payload_2yfbdmwa/ansible_community.general.gitlab_hook_payload.zip/ansible_collections/community/general/plugins/modules/source_control/gitlab/gitlab_hook.py\", line 260, in update_hook
  File \"/Users/k-petrov/Library/Python/3.8/lib/python/site-packages/gitlab/base.py\", line 142, in __getattr__
    raise AttributeError(message) from exc
AttributeError: ProjectHook object has no attribute token
```

gitlab API doesn`t return token (https://docs.gitlab.com/ee/api/projects.html#get-project-hook)

ipython:
```
In [51]: hook.attributes.keys()
Out[51]: dict_keys(['id', 'url', 'created_at', 'push_events', 'tag_push_events', 'merge_requests_events', 'repository_update_events', 'enable_ssl_verification', 'project_id', 'issues_events', 'confidential_issues_events', 'note_events', 'confidential_note_events', 'pipeline_events', 'wiki_page_events', 'deployment_events', 'job_events', 'releases_events', 'push_events_branch_filter'])
```

### CODE REVIEW

File: plugins/modules/source_control/gitlab/gitlab_hook.py [hunk 1]
```diff
@@ -257,7 +257,7 @@ def update_hook(self, hook, arguments):
 
         for arg_key, arg_value in arguments.items():
             if arguments[arg_key] is not None:
-                if getattr(hook, arg_key) != arguments[arg_key]:
+                if getattr(hook, arg_key, '') != arguments[arg_key]:
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_hook.py:260]: Doesn't `None` make more sense than `''`?
```suggestion
                if getattr(hook, arg_key, None) != arguments[arg_key]:
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! Could you please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment)?

File: plugins/modules/source_control/gitlab/gitlab_hook.py (hunk 1 above)
Comment (crashwind) [reply plugins/modules/source_control/gitlab/gitlab_hook.py]: fixed

File: changelogs/fragments/4668-gitlab_hook-use-None-for-non-existent-attr.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - gitlab_hook - use None for non-existent attribute, e.g. 'token'
```
Comment (felixfontein) [inline changelogs/fragments/4668-gitlab_hook-use-None-for-non-existent-attr.yml:2]:
```suggestion
  - gitlab_hook - avoid errors during idempoency check when an attribute does not exist (https://github.com/ansible-collections/community.general/pull/4668).
```

**Comment by felixfontein** [comment]
@Lunik @Shaps @dj-wasabi @lgatellier @marwatk @metanovii @nejch @scodeman @sh0shin @suukit @waheedi @zanssa would be glad if someone could take a look at this change.

**Comment by lgatellier** [comment]
Seems OK for me.

But this raises a question :
We configure a GitLab webhook with a token, and there is no GitLab API which returns this token, so we're not able to determine if the token has been changed.

Thus, when a token is present, the ansible task will always terminate with `changed` status.
@felixfontein in general, is it better to have a task which is always `changed`, or always `ok` without knowing if anything changed ?

**Comment by felixfontein** [comment]
@lgatellier good question. It's probably best to always return `changed` when the user specified a token, and not changed when the user does not (assuming nothing else changed).

**Comment by crashwind** [comment]
@felixfontein always return `changed` if token defined in task by this request as you described above

File: changelogs/fragments/4668-gitlab_hook-use-None-for-non-existent-attr.yml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - gitlab_hook - avoid errors during idempoency check when an attribute does not exist (https://github.com/ansible-collections/community.general/pull/4668).
```
Comment (felixfontein) [inline changelogs/fragments/4668-gitlab_hook-use-None-for-non-existent-attr.yml:2]:
```suggestion
  - gitlab_hook - avoid errors during idempotency check when an attribute does not exist (https://github.com/ansible-collections/community.general/pull/4668).
```

**Comment by felixfontein** [comment]
@lgatellier do I assume correctly from your comment and the thumbs up that this can be merged? If not please tell me until tomorrow morning, otherwise it will be too late :)

**Review by felixfontein — APPROVED** [review]

**Comment by lgatellier** [comment]
> @lgatellier do I assume correctly from your comment and the thumbs up that this can be merged? If not please tell me until tomorrow morning, otherwise it will be too late :)

You assume CORRECTLY 😉

LGTM !

**Comment by felixfontein** [comment]
@crashwind thanks for fixing this!
@lgatellier thanks for reviewing!

**Pull request merged**